### PR TITLE
Lexicographic serializer

### PR DESCRIPTION
**Pull request by sjwiesman**

### CODE REVIEW

File: flink-core/src/main/java/org/apache/flink/api/common/typeutils/base/LexicographicLongSerializer.java [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+        target.write(source, 8);
+    }
+
+    @Override
+    public TypeSerializerSnapshot<Long> snapshotConfiguration() {
+        return new LexicographicLongSerializerSnapshot();
+    }
+
+    /** Serializer configuration snapshot for compatibility and format evolution. */
+    @SuppressWarnings("WeakerAccess")
```
Comment (NicoK) [inline flink-core/src/main/java/org/apache/flink/api/common/typeutils/base/LexicographicLongSerializer.java:110]: nit: I see this `@SuppressWarnings("WeakerAccess")` in numerous implementations of `SimpleTypeSerializerSnapshot` but I'm actually not sure it is really needed, is it?

File: flink-core/src/test/java/org/apache/flink/api/common/typeutils/base/LexicographicLongSerializerTest.java [hunk 2]
```diff
@@ -0,0 +1,169 @@
...
+        }
+
+        @Override
+        public void skipBytesToRead(int numBytes) throws IOException {
+            while (numBytes > 0) {
+                int skipped = skipBytes(numBytes);
+                numBytes -= skipped;
+            }
+        }
+    }
```
Comment (NicoK) [inline flink-core/src/test/java/org/apache/flink/api/common/typeutils/base/LexicographicLongSerializerTest.java:168]: This is already available in `SerializerTestBase` - let's reuse this as much as possible (I'll propose a patch for that)

File: flink-examples/flink-examples-streaming/src/main/java/org/apache/flink/streaming/examples/temporal/TemporalJoin.java [hunk 3]
```diff
@@ -0,0 +1,156 @@
...
+
+        @Override
+        public void onTimer(
+                long timestamp,
+                KeyedCoProcessFunction<String, Transaction, CurrencyRate, Joined>.OnTimerContext
+                        ctx,
+                Collector<Joined> out)
+                throws Exception {
+
+            TemporalValue<Double> rate = this.rate.valueAtOrBefore(timestamp);
```
Comment (NicoK) [inline flink-examples/flink-examples-streaming/src/main/java/org/apache/flink/streaming/examples/temporal/TemporalJoin.java:107]: I believe, we'll have to handle the not found / `null` case here as well...

File: flink-examples/flink-examples-streaming/src/main/java/org/apache/flink/streaming/examples/temporal/TemporalJoin.java [hunk 4]
```diff
@@ -0,0 +1,156 @@
...
+        }
+
+        @Override
+        public void processElement2(
+                CurrencyRate value,
+                KeyedCoProcessFunction<String, Transaction, CurrencyRate, Joined>.Context ctx,
+                Collector<Joined> out)
+                throws Exception {
+
+            rate.update(ctx.timestamp(), value.rate);
```
Comment (NicoK) [inline flink-examples/flink-examples-streaming/src/main/java/org/apache/flink/streaming/examples/temporal/TemporalJoin.java:96]: Ideally, we only keep the most-recent rate before the last watermark and ignore the rest (and also clean up based on that). Now the current code only cleans up if there are transactions which may become a problem.
I can think of two ways to solve this:
1. clean up here by trying to find rates before the current watermark and deleting all but the most recent one
2. clean up with a timer for the next watermark
3. clean up regularly

I'm not sure which of these options is best - each of them have performance implications:
1. has 2 additional lookups for every new rate (`valueAtOrBefore(current_watermark)` and then finding more before this one
2. adds quite a few more timers
3. collects unused rates for a while

Maybe (3) is the best, especially if combined with the cleanup at the transaction timers?

File: flink-runtime/src/main/java/org/apache/flink/runtime/state/DefaultKeyedStateStore.java [hunk 5]
```diff
@@ -36,6 +36,7 @@
 import org.apache.flink.api.common.state.ValueState;
 import org.apache.flink.api.common.state.ValueStateDescriptor;
 import org.apache.flink.api.common.typeutils.TypeSerializer;
+import org.apache.flink.api.common.typeutils.base.LexicographicLongSerializer;
 import org.apache.flink.api.common.typeutils.base.LongSerializer;
```
Comment (NicoK) [inline flink-runtime/src/main/java/org/apache/flink/runtime/state/DefaultKeyedStateStore.java:40]:
```suggestion
```

File: flink-runtime/src/main/java/org/apache/flink/runtime/state/UserFacingTemporalListState.java [hunk 6]
```diff
@@ -33,18 +36,45 @@
...
+    public TemporalValue<Iterable<T>> getAtOrAfter(long time) throws Exception {
+        TemporalNavigator navigator = Preconditions.checkNotNull(internalState.getNavigator());
+        Long found = navigator.atOrAfter(time);
+        if (found == null) {
+            return new TemporalValue<>(Long.MIN_VALUE, emptyState);
+        }
+
+        internalState.setCurrentNamespace(found);
+        Iterable<T> original = originalState.get();
+        Iterable<T> iter = original != null ? original : emptyState;
```
Comment (NicoK) [inline flink-runtime/src/main/java/org/apache/flink/runtime/state/UserFacingTemporalListState.java:74]: I think, I'd like to extend the current functionality to retrieve key+value at once (from the navigator) - I'll prepare a separate PR for this and we can see about the practical difference)

File: flink-runtime/src/main/java/org/apache/flink/runtime/state/internal/InternalKvState.java [hunk 7]
```diff
@@ -20,9 +20,13 @@
 
 import org.apache.flink.api.common.state.State;
 import org.apache.flink.api.common.typeutils.TypeSerializer;
+import org.apache.flink.api.common.typeutils.base.LexicographicLongSerializer;
```
Comment (NicoK) [inline flink-runtime/src/main/java/org/apache/flink/runtime/state/internal/InternalKvState.java:23]:
```suggestion
```

File: flink-runtime/src/main/java/org/apache/flink/runtime/state/internal/InternalKvState.java [hunk 8]
```diff
@@ -20,9 +20,13 @@
...
 import org.apache.flink.runtime.state.StateEntry;
 
+import javax.annotation.Nullable;
+
 import java.util.Collection;
+import java.util.Optional;
```
Comment (NicoK) [inline flink-runtime/src/main/java/org/apache/flink/runtime/state/internal/InternalKvState.java:29]:
```suggestion
```

File: flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/operators/join/temporal/TemporalRowTimeJoinOperator.java [hunk 9]
```diff
@@ -150,6 +157,12 @@ public void open() throws Exception {
         joinCondition.setRuntimeContext(getRuntimeContext());
         joinCondition.open(new Configuration());
 
+        tLeftState = getRuntimeContext()
+                .getTemporalListState(new ListStateDescriptor<>(LEFT_STATE_NAME, leftType));
+
+        tRightState = getRuntimeContext()
+                .getTemporalState(new ValueStateDescriptor<>(RIGHT_STATE_NAME, rightType));
+
```
Comment (NicoK) [inline flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/operators/join/temporal/TemporalRowTimeJoinOperator.java:165]: I think this change is incomplete, it's unit test in `TemporalRowTimeJoinOperatorTest` is also failing...

**Comment by NicoK** [comment]
FYI: I'll mark comments above with :+1: when I have a commit in my branch to fix them

File: flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/operators/join/temporal/TemporalRowTimeJoinOperator.java (hunk 9 above)
Comment (sjwiesman) [reply flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/operators/join/temporal/TemporalRowTimeJoinOperator.java:165]: Oh, it’s not. I didn’t mean to commit that file

Comment (NicoK) [reply flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/operators/join/temporal/TemporalRowTimeJoinOperator.java:165]: got that covered in https://github.com/dataArtisans/flink/pull/13

**Pull request merged**